Support Document Symbols

mattmasson opened the pull request:
Adds support for the [Document Symbols LSP](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_documentSymbol) functionality.

In VS Code, this shows up in the Outline window like this:

![image](https://user-images.githubusercontent.com/5509937/85796339-c4a4a500-b707-11ea-8a50-6865d6901cc5.png)

Supports:

- Section members
- Let expression members
- Record fields

Also sets the correct symbol type when constants/literals are used.

Further improvements could be:

- Support (partial outline) for members with syntax errors (i.e. context nodes)
- Include function signature 
- Tracing for error cases
- Lookup appropriate type info for library constants and literals
- Recognize correct type for `Value.ReplaceType` (causes functions to appear as variables)

JordanBoltonMN commented on src/language-services/documentSymbols.ts:
```diff
@@ -0,0 +1,166 @@
+        PQP.Traverse.VisitNodeStrategy.BreadthFirst,
+        visitNode,
+        PQP.Traverse.expectExpandAllAstChildren,
+        undefined,
```
Can you add a function here? It can hugely speed up a traversal. Looks like anything that can't be/contain either a Section/Record/IdentifierPairedExpression can be ignored.

mattmasson replied:
I added the function, but had a hard time figuring out where to stop the traversal.

JordanBoltonMN commented on src/language-services/documentSymbols.ts:
```diff
@@ -0,0 +1,166 @@
+        },
+    };
+
+    return PQP.Traverse.tryTraverseAst(
```
Using `tryTraverseAst` limits this to a fully parsed document. Changing it to `tryTraverseXor` would obviously make the implementation tricker (though that's mitigated by a bunch of already made helper functions), but it would work in the most common scenario.

JordanBoltonMN commented on src/language-services/inspectionUtils.ts:
```diff
@@ -92,6 +95,24 @@ export function getSymbolsForLetExpression(expressionNode: PQP.Language.Ast.LetE
+export function getSymbolsForRecord(
+    recordNode: PQP.Language.Ast.RecordExpression | PQP.Language.Ast.RecordLiteral,
+): DocumentSymbol[] {
+    const documentSymbols: DocumentSymbol[] = [];
```
nit: this could be replaced with a `recordNode.content.elements.reduce(...)`. It's arguable if that makes it more/less readable.

mattmasson commented:
I removed the top level Section (Module) symbol, since it didn't really add anything to the symbols functionality (i.e. it's always the top of the document anyways).

I switched to `TryTraverseXor`, but bail on context nodes. I think the right thing to do is return a node with a best guess of its required properties (name, kind, range) - that allows errors to show up in the `Outline` window in VS Code. But I'd like to see how useful we actually find the feature before putting too much effort into this.
